gradio/chat_interface.py

```diff
@@ -0,0 +1,269 @@
+            self._display_input,
+            [self.saved_input, self.chatbot],
+            [self.chatbot],
+            api_name=False,
+            queue=False,
+        ).then(
+            submit_fn,
+            [self.saved_input, self.chatbot],
+            [self.chatbot],
+            api_name=False,
+        )
+
+        if self.submit_btn:
+            self.submit_btn.click(
+                self._clear_and_save_textbox,
+                [self.textbox],
+                [self.textbox, self.saved_input],
+                api_name=False,
+                queue=False,
+            ).then(
+                self._display_input,
+                [self.saved_input, self.chatbot],
+                [self.chatbot],
+                api_name=False,
+                queue=False,
+            ).then(
+                submit_fn,
+                [self.saved_input, self.chatbot],
+                [self.chatbot],
+                api_name=False,
+            )
+        
+        if self.retry_btn:
+            self.retry_btn.click(
+                self._delete_prev_fn,
+                [self.chatbot],
+                [self.chatbot, self.saved_input],
+                api_name=False,
+                queue=False,
+            ).then(
+                self._display_input,
+                [self.saved_input, self.chatbot],
+                [self.chatbot],
+                api_name=False,
+                queue=False,
+            ).then(
+                submit_fn,
+                [self.saved_input, self.chatbot],
+                [self.chatbot],
+                api_name=False,
+            )
+
+        if self.delete_last_btn:
+            self.delete_last_btn.click(
+                self._delete_prev_fn,
+                [self.chatbot],
+                [self.chatbot, self.saved_input],
+                api_name=False,
+                queue=False,
+            )
+
+        if self.clear_btn:
+            self.clear_btn.click(
+                lambda :([], None), 
+                None, 
+                [self.chatbot, self.saved_input],
+                queue=False,
+                api_name=False)
+            
+    def _setup_api(self):
+        if inspect.isgeneratorfunction(self.fn):
+            api_fn = self._api_stream_fn
+        else:
+            api_fn = self._api_submit_fn
+
+        # Use a gr.State() instead of self.chatbot so that the API doesn't require passing forth
+        # a chat history, instead it is just stored internally in the state.
+        history = State([])
+
+        self.fake_api_btn.click(
+            api_fn,
+            [self.textbox, history],
+            [self.textbox, history],
+            api_name="chat",
+        )
+
+    def _clear_and_save_textbox(self, message):
+        return "", message
+
+    def _display_input(self, message: str, history: list[list[str]]):
+        history.append((message, None))
+        return history
+
+    def _submit_fn(self, message: str, history: list[list[str]]):
```

Comment:
So the history contains the user message (since we call _display_history)? I think that's confusing. I would expect the history to only contain the history up to but not including the current message.

Reply:
Yeah its confusing since In `_display_input`, `history` contains the history not including the current message, but by the time that `_submit_fn` is called, `history` includes the current message. These are all internal methods so I don't think it matters too much, but I'll rename `history` to be `history_with_input` or something

Reply:
I'm worried this will introduce bugs into demos built with ChatInterface. I think it would be natural for a developer to do something like the following (pseudocode):

```python
def chat(message, history):
    chain.history = history
    return chain.predict(message)
```
And then pass duplicate entries for `message` to the chain. BTW I think there is an inconsistency between `_api_stream_fn` and `_api_fn` the latter does not include the latest message in the chain but the former does.

Whatever we pick we should be consistent but I think it's more natural for the history to not include the latest message.

Reply:
Oh wait I made a mistake in `_submit_fn`. The history should absolutely NOT include the last message. I misunderstood what you were saying earlier. Sorry, I'll fix and test this out